### PR TITLE
feat: add node-mirror:<releaseDir> options

### PR DESCRIPTION
**Pull request by gluxon**
fixes: #4039

### CODE REVIEW

File: packages/plugin-commands-env/src/constants.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export const DEFAULT_NODE_MIRROR = 'https://nodejs.org/download'
```
Comment (gluxon) [inline packages/plugin-commands-env/src/constants.ts]: I don't see many `constants` files in the repo. Is this something we want to discourage? I moved this here to make it more consistently shared between `resolveNodeVersion` and `installNode`.

File: packages/plugin-commands-env/src/node.ts [hunk 2]
```diff
@@ -115,14 +120,14 @@ async function downloadAndUnpackZip (
...
-function getNodeJSTarball (nodeVersion: string, releaseDir: string) {
+function getNodeJSTarball (nodeVersion: string, releaseDir: string, nodeMirror: string) {
   const platform = process.platform === 'win32' ? 'win' : process.platform
   const arch = normalizeArch(process.platform, process.arch)
   const extension = platform === 'win' ? 'zip' : 'tar.gz'
   const pkgName = `node-v${nodeVersion}-${platform}-${arch}`
   return {
     pkgName,
-    tarball: `https://nodejs.org/download/${releaseDir}/v${nodeVersion}/${pkgName}.${extension}`,
+    tarball: `${nodeMirror}/${releaseDir}/v${nodeVersion}/${pkgName}.${extension}`,
```
Comment (gluxon) [inline packages/plugin-commands-env/src/node.ts]: Do we want to validate the URI in some way? I think someone could set this to a `file:///` URI in theory. That seems fine to me, but maybe there's security ramifications I'm forgetting.

File: packages/plugin-commands-env/src/resolveNodeVersion.ts [hunk 3]
```diff
@@ -1,15 +1,16 @@
...
 interface NodeVersion {
   version: string
   lts: false | string
 }
 
-export default async function resolveNodeVersion (fetch: FetchFromRegistry, rawVersionSelector: string) {
+export default async function resolveNodeVersion (fetch: FetchFromRegistry, rawVersionSelector: string, nodeMirror: string = DEFAULT_NODE_MIRROR) {
   const { releaseDir, version } = parseNodeVersionSelector(rawVersionSelector)
-  const response = await fetch(`https://nodejs.org/download/${releaseDir}/index.json`)
+  const response = await fetch(`${nodeMirror}/${releaseDir}/index.json`)
```
Comment (gluxon) [inline packages/plugin-commands-env/src/resolveNodeVersion.ts]: This PR doesn't currently test that the `resolveNodeVersion` function actually uses the new `node-mirror` option. This was a bit difficult to intercept without setting up Jest mocks for `@pnpm/fetch` or larger refactors.

Happy to do a `jest.mock("pnpm/fetch")` test to spy this line if that's desirable. For now I thought that was more complicated than it was worth.

File: .changeset/kind-toys-relate.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
+"@pnpm/config": minor
+"@pnpm/plugin-commands-env": minor
+---
```
Comment (zkochan) [inline .changeset/kind-toys-relate.md:5]: add also pnpm as minor

File: packages/plugin-commands-env/src/node.ts (hunk 2 above)
Comment (zkochan) [inline packages/plugin-commands-env/src/node.ts]: what if nodeMirror will be set with a trailing slash?

Comment (gluxon) [reply packages/plugin-commands-env/src/node.ts]: Happy to normalize this. Does tacking it on if it's missing seem enough?

```ts
function normalizeNodeMirrorOption(nodeMirror: string) {
  return nodeMirror.endsWith("/") ? nodeMirror : `${nodeMirror}/`;
}
```

Comment (zkochan) [reply packages/plugin-commands-env/src/node.ts]: yes

File: packages/plugin-commands-env/src/constants.ts (hunk 1 above)
Comment (zkochan) [reply packages/plugin-commands-env/src/constants.ts]: it is ok

File: packages/plugin-commands-env/src/node.ts (hunk 2 above)
Comment (zkochan) [reply packages/plugin-commands-env/src/node.ts]: I don't think there's a need to validate the protocol

File: packages/plugin-commands-env/src/resolveNodeVersion.ts (hunk 3 above)
Comment (zkochan) [reply packages/plugin-commands-env/src/resolveNodeVersion.ts]: Yes, I think it should be tested. You may use a mock or nock. You can see how nock is used here: https://github.com/pnpm/pnpm/blob/main/packages/npm-resolver/test/index.ts

Comment (gluxon) [reply packages/plugin-commands-env/src/resolveNodeVersion.ts]: Woah, I didn't know about `nock`. That makes this so easy. Thank you for telling me about it and linking to an example! 🙂

**Comment by gluxon** [comment]
@zkochan One aspect of this is that the new `node-mirror` option will actually be a bit difficult to use at my org. The existing mirror currently follows `NVM_NODEJS_ORG_MIRROR` behavior. In other words, it's a mirror to https://nodejs.org/dist, which I believe is identical to the default https://nodejs.org/download/release path pnpm uses.

The problem is that our internal `NVM_NODEJS_ORG_MIRROR` style mirror doesn't end with `/release`. We can proxy something from `/release` to the existing mirror, but was curious if you had any ideas to make this easier to use. From looking at the `use-node-version` implementation here though, that didn't look easy.

**Comment by zkochan** [comment]
Make it a pattern. So that you'll have to do something like:

```
node-mirror=https://foo.bar/download/<RELEASE_DIR>/
```

**Comment by zkochan** [comment]
maybe someone else has alternative suggestions. cc @pnpm/collaborators

**Comment by gluxon** [comment]
Ah, so then `DEFAULT_NODE_MIRROR` becomes the following?

```
export const DEFAULT_NODE_MIRROR = 'https://nodejs.org/download/<RELEASE_DIR>/'
```

**Comment by zkochan** [comment]
yes

**Comment by gluxon** [comment]
The `<RELEASE_DIR>` pattern works for me. Should we wait a bit longer for a response from @pnpm/collaborators anyone on that list has thoughts?

**Comment by sodatea** [comment]
The npm mirror in China looks like this:
- https://npmmirror.com/mirrors/node/
- https://npmmirror.com/mirrors/node-rc/
- https://npmmirror.com/mirrors/node-nightly/

How does pnpm support this pattern?

**Comment by zkochan** [comment]
ok, as an alternative, we can add something like

```
node-mirror=https://npmmirror.com/mirrors/node/
node-mirror:rc=https://npmmirror.com/mirrors/node-rc/
node-mirror:nightly=https://npmmirror.com/mirrors/node-nightly/
```

**Comment by gluxon** [comment]
I like that option. For that first line, would we perhaps want `:release` to be explicit?

```ini
node-mirror:release=https://npmmirror.com/mirrors/node/
```

I could see some people being surprised that just `node-mirror` doesn't apply when setting an `rc` or `nightly` version maybe.

**Comment by zkochan** [comment]
sounds good. If you want to ready this options, use the rawConfig option

**Comment by gluxon** [comment]
I'll give this approach a shot. Thanks!

**Review by zkochan — APPROVED** [review]

**Comment by gluxon** [comment]
Thanks a lot @zkochan!

**Comment by zkochan** [comment]
Thanks for your contribution!

**Pull request merged**